smack-integration-test/src/main/java/org/jivesoftware/smackx/muc/MultiUserChatRolesAffiliationsPrivilegesIntegrationTest.java

```diff
@@ -619,4 +621,295 @@ public void mucPresenceTestForWitnessingKick() throws Exception {
+        resultSyncPoint.waitForResult(timeout);
+
+        try {
+            assertEquals(mucAsSeenByOne.getOccupantsCount(), 3);
+            assertEquals(MUCRole.moderator, mucAsSeenByOne.getOccupant(
+                            JidCreate.entityFullFrom(mucAddress + "/" + nicknameOne.toString())).getRole());
+            assertEquals(MUCRole.moderator, mucAsSeenByOne.getOccupant(
+                            JidCreate.entityFullFrom(mucAddress + "/" + nicknameTwo.toString())).getRole());
+            assertEquals(MUCRole.participant, mucAsSeenByOne.getOccupant(
+                            JidCreate.entityFullFrom(mucAddress + "/" + nicknameThree.toString())).getRole());
+        } finally {
+            tryDestroy(mucAsSeenByOne);
+        }
+    }
+
+    /**
+     * Asserts that a moderated room assigns the correct default roles for a given affiliation
+     *
+     * <p>From XEP-0045 § 5.1.2:</p>
+     * <blockquote>
+     * ...the initial default roles that a service SHOULD set based on the user's affiliation...
+     * </blockquote>
+     *
+     * @throws Exception when errors occur
+     */
+    @SmackIntegrationTest
+    public void mucTestDefaultRoleForAffiliationInModeratedRoom() throws Exception {
+        EntityBareJid mucAddress = getRandomRoom("smack-inttest-moderatedroles");
+
+        MultiUserChat mucAsSeenByOne = mucManagerOne.getMultiUserChat(mucAddress);
+        MultiUserChat mucAsSeenByTwo = mucManagerTwo.getMultiUserChat(mucAddress);
+        MultiUserChat mucAsSeenByThree = mucManagerThree.getMultiUserChat(mucAddress);
+
+        final Resourcepart nicknameOne = Resourcepart.from("one-" + randomString);
+        final Resourcepart nicknameTwo = Resourcepart.from("two-" + randomString);
+        final Resourcepart nicknameThree = Resourcepart.from("three-" + randomString);
+
+        createModeratedMuc(mucAsSeenByOne, nicknameOne.toString());
+        mucAsSeenByTwo.join(nicknameTwo);
+        mucAsSeenByThree.join(nicknameThree);
+
+        final ResultSyncPoint<String, Exception> resultSyncPoint = new ResultSyncPoint<>();
+        mucAsSeenByOne.addParticipantStatusListener(new ParticipantStatusListener() {
+            @Override
+            public void adminGranted(EntityFullJid participant) {
+                resultSyncPoint.signal("done");
+            }
+        });
+        mucAsSeenByOne.grantAdmin(conTwo.getUser().asBareJid());
+        resultSyncPoint.waitForResult(timeout);
+
+        try {
+            assertEquals(mucAsSeenByOne.getOccupantsCount(), 3);
+            assertEquals(MUCRole.moderator, mucAsSeenByOne.getOccupant(
+                            JidCreate.entityFullFrom(mucAddress + "/" + nicknameOne.toString())).getRole());
+            assertEquals(MUCRole.moderator, mucAsSeenByOne.getOccupant(
+                            JidCreate.entityFullFrom(mucAddress + "/" + nicknameTwo.toString())).getRole());
+            assertEquals(MUCRole.visitor, mucAsSeenByOne.getOccupant(
```

Comment:
This is the only test that fails for me on ejabberd, as the role is 'participant' not 'visitor'. Had not yet the time to look into it.

Reply:
In this implementation, the entity with nickname `nicknameThree` has joined a moderated room. 

Unlike the other two entities (`nicknameOne` created the room, `nicknameTwo` was granted Admin), no affiliation changes are applied to `nicknameThree`. The entity therefore doesn't have an affiliation.

According to [section 5.1.2, table 4](https://xmpp.org/extensions/xep-0045.html#table-4), the default role of such an entity SHOULD be `Visitor`. To me, it seems that this test asserts the correct status.

`Participant` would be appropriate for a similar scenario that plays out in an unmoderated room. That scenario is tested in a different test, `mucTestDefaultRoleForAffiliationInUnmoderatedRoom` (just above the one that we're discussing here).

Of note is that these tests assert a SHOULD definition. It's debatable if it warrants a test failure.

Reply:
I also think that this test asserts a SHOULD correctly. However ejabberd I my primary test platform. One way to move forward is to have a `static void setModeratedRoomDefaultRole(MUCRole mucRole)` method and a `static MUCRole moderedRoomDefaultRole = MUCRole.participant` field.

Reply:
How about exposing a sinttest setting for this, so that you can override it for ejabberd?

Reply:
I would usually prefer default compatibility with ejabberd than openfire. Simply because I use ejabberd as my primary and integration test server. However, I may be talked into defaulting to the XEPs SHOULD, if you create a new sinttest configuration setting that I can put into my `~/.config/smack-integration-test/properties`, e.g.

```
compatibilityMode=ejabberd
```

Reply:
How do you feel about this? Feel more horizontally scalable:

```
    @SmackIntegrationTest
    public void mucTestDefaultRoleForAffiliationInUnmoderatedRoom() throws Exception {
        if(!environment.configuration.enableNonMandatoryXepChecks){
            return; //Don't test this
        }
```

Reply:
I also prefer this as it's more deterministic. A third XMPP server may have a third behaviour, and a fourth might have a fourth behaviour. Instead, this either enables or disables testing against the XEP's recommended behaviour.

Reply:
But... should this be a TestNotPossibleException? It seems to fit, except that it's not "NotPossible" so much as it is "NotDesirable" :)